src/App.tsx

```diff
@@ -1,223 +1,216 @@
+    [filterBy, todos],
+  );
 
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
+    if (user) {
+      setErrMsg('');
+      getTodos(user.id)
+        .then(setTodos)
+        .catch(setErrMsg);
+    }
+  }, []);
+
+  const handleErrMsg = useCallback((titileErr: string) => {
+    setErrMsg(titileErr);
+
+    setTimeout(() => setErrMsg(''), 3000);
+  }, []);
+
+  const handleCreateNewTodo = useCallback(() => {
+    if (!newTodoTitle) {
+      handleErrMsg('Title can\'t be empty');
+
+      return;
     }
+
+    const newTodo = {
+      title: newTodoTitle,
+      userId: user?.id as number,
+      completed: false,
+    };
+
+    setErrMsg('');
+    setIsLoadingTodos(prev => [...prev, -1]);
+    setTodos(prevTodos => [...prevTodos, { ...newTodo, id: -1 }]);
+    createNewTodo(newTodo)
+      .then(res => {
+        setNewTodoTitle('');
+        setTodos(prevTodos => prevTodos.map(todo => {
+          if (todo.id === -1) {
+            return { ...res as Todo };
+          }
+
+          return todo;
+        }));
+      })
+      .catch(() => {
+        handleErrMsg('Unable to add a todo');
+        setTodos(prevTodos => prevTodos.slice(0, -1));
+      })
+      .finally(() => setIsLoadingTodos(prev => prev.filter(n => n !== -1)));
+  }, [newTodoTitle, user]);
+
+  const handleRemoveTodo = useCallback((id: number) => {
+    setErrMsg('');
+    setIsLoadingTodos(prev => [...prev, id]);
+    removeTodo(id)
+      .then((res) => {
+        if (!res) {
+          throw new Error();
+        }
+
+        setTodos(prevTodos => prevTodos.filter(todo => todo.id !== id));
+      })
+      .catch(() => handleErrMsg('Unable to delete a todo'))
+      .finally(() => setIsLoadingTodos(prev => prev.filter(n => n !== id)));
   }, []);
 
+  const handleChangeTodo = useCallback((id: number, data:{}) => {
+    setErrMsg('');
+    setIsLoadingTodos(prev => [...prev, id]);
+    updateTodo(id, data)
+      .then(res => {
+        setIsLoadingTodos(prev => [...prev, id]);
+        setTodos(prevTodos => prevTodos.map(todo => {
+          if (todo.id === id) {
+            return { ...res as Todo };
+          }
+
+          return todo;
+        }));
+      })
+      .catch(() => handleErrMsg('Unable to update a todo'))
+      .finally(() => setIsLoadingTodos(prev => prev.filter(n => n !== id)));
+  }, []);
+
+  const completeAll = useCallback(() => {
+    if (todos.every(({ completed }) => completed)) {
+      todos.forEach(
+        ({ id }) => handleChangeTodo(id, { completed: false }),
+      );
+    } else {
+      todos.forEach(({ completed, id }) => {
+        if (!completed) {
+          handleChangeTodo(id, { completed: true });
+        }
+      });
+    }
+  }, [todos]);
+
+  const clearCompleted = useCallback(() => {
+    todos.forEach(({ completed, id }) => {
+      if (completed) {
+        handleRemoveTodo(id);
+      }
+    });
+  }, [todos]);
```

Comment:
let's use `Promise.all` here, since todos should be removed simultaneously

<img width="628" alt="image" src="https://user-images.githubusercontent.com/55802630/187189271-cc2488a1-b497-4d58-bdfd-e6c122b86282.png">

Reply:
My deletions work like several separate deletions and run at the same time, their deletion ends at different times only. I don't agree with you about Promise.all, in this case it's wrong to use it.

Reply:
I can prove why

Reply:
I think we need only compute requests simultaneously, but not end deletion at the same time

Reply:
I only doubt about `Promise.all`
Now i think `Promise.allSettled` is more appropriate here

Reply:
>in this case it's wrong to use it. I can prove why

but I'll be glad to hear why it is wrong and will use it later if so